src/runtime/thread_pool.cc

```diff
@@ -127,99 +130,138 @@ class ParallelLauncher {
+      std::unique_lock<std::mutex> lock(mutex_);
       cv_.notify_one();
     }
   }
+
   /*!
-   * \brief Pop task from the queue
-   * \param task The task to be poped.
-   * \param timeout The number of cycles to spin before sleep.
-   * \return Whether pop is successful or we need to exit now.
+   * \brief Pop a task out of the queue and condition wait if no tasks.
+   * \param output The pointer to the task to be dequeued.
+   * \param timeout The number of iterations to spin before sleep.
+   * \return Whether pop is successful (true) or we need to exit now (false).
    */
-  bool Pop(Task* task, int timeout = 10) {
-    std::unique_lock<std::mutex> lock(mutex_);
-    if (num_pending_ != 0) {
-      *task = ring_[head_];
-      head_ = (head_ + 1) % ring_.size();
-      --num_pending_;
-      if (exit_now_.load()) return false;
-    } else {
-      lock.unlock();
-      // do a bit spin and busy waiting before sleep.
-      for (int i = 0; i < timeout && num_pending_ == 0; ++i) {
+  bool Pop(Task* output, uint32_t timeout = 100000) {
+    // busy wait a bit when the queue is empty
+    // if a new task comes to the queue quickly, this wait avoid the worker from sleeping
+    for (uint32_t i = 0; i < timeout && pending_.load() == 0; ++i) {
         std::this_thread::yield();
       }
-      lock.lock();
-      ++nwait_consumer_;
+    if (pending_.fetch_sub(1) == 0) {
+      std::unique_lock<std::mutex> lock(mutex_);
       cv_.wait(lock, [this] {
-          return num_pending_ != 0 || exit_now_.load();
+          return pending_.load() >= 0 || exit_now_.load();
         });
-      --nwait_consumer_;
-      *task = ring_[head_];
-      head_ = (head_ + 1) % ring_.size();
-      --num_pending_;
-      if (exit_now_.load()) return false;
     }
-    return true;
+    return Dequeue(output);
+    }
+
+  /*!
+   * \brief Signal to terminate the worker.
+   */
+  void SignalForKill() {
+    exit_now_.store(true);
+    cv_.notify_all();
   }
 
  private:
-  // Number of the elments in the queue
-  uint32_t num_pending_{0};
-  // Queue head
-  uint32_t head_{0};
-  // Number of consumers to wait.
-  uint32_t nwait_consumer_{0};
+  struct node_t {
+    node_t* next;
+    Task data;
+  };
+
+  typedef std::aligned_storage<sizeof(Task), std::alignment_of<Task>::value>::type \
+    aligned_t;
+
+  typedef char cache_line_pad_t[64];
+  cache_line_pad_t _pad0;
+  // size of the queue, the queue can host size_ - 1 items at most
+  const uint8_t size_;
+  // pointer to access the item
+  Task* const buffer_;
+
+  cache_line_pad_t _pad1;
+  // queue head, where one gets a task from the queue
+  std::atomic<uint8_t> head_;
+
+  cache_line_pad_t _pad2;
+  // queue tail, when one puts a task to the queue
+  std::atomic<uint8_t> tail_;
+
+  cache_line_pad_t _pad3;
+  // pending tasks in the queue
+  std::atomic<int8_t> pending_{0};
+
+  cache_line_pad_t _pad4;
+  // signal for exit now
+  std::atomic<bool> exit_now_{false};
+
+  cache_line_pad_t _pad5;
```

Comment:
no padding need in here since the lock is slow path and the owner of exit_now is the same as cv